### PR TITLE
Support batch sessions

### PR DESCRIPTION
**Pull request by JustEdro**
Hey there! 
I've created and somehow tested the batch endpoint, but I'm not familiar with the test system you use in this project, could you please help me to add some?

### CODE REVIEW

**Comment by acroz** [comment]
Hi, thanks for the contribution!

At the moment, there are two types of tests:

- Unit tests of the client and models (tests/test_client.py and tests/test_models.py)
- Integration tests of the `LivySession` helper (tests/test_integration.py)

Tests are run using [pytest](https://pytest.readthedocs.io/en/latest/) -- you might want to take a look at some minimal examples in their documentation first. I suggest running the tests with [tox](https://tox.readthedocs.io/en/latest/), which will install additional dependencies required:

```sh
tox -e py36
```

Replace `py36` with `py37` to run on Python 3.7 instead of Python 3.6.

By default, this will _not_ run the integration tests, as they are slower and require a running Livy server to test against. To run integration tests, you can run a Livy server with Docker:

```sh
docker run --publish 8998:8998 acroz/livy
```

Then, in a separate shell, run the integration tests with:

```sh
tox -e py36-integration
```

You can also run linting checks and type checks with the following:

```sh
tox -e flake8
tox -e black
tox -e mypy
```

## Adding tests

I suggest starting with a test for the new `Batch` model you've added, in particular testing its `from_json` method in a similar way to the existing tests in tests/test_models.py.

After that, I'd suggest adding tests to tests/test_client.py to test the methods you've added to `LivyClient`. Those tests use the requests_mock to mock out the actual HTTP calls so it's possible to test the client logic without an actual livy server running. These tests also mock out the `from_json` (e.g. `mocker.patch.object(Statement, "from_json")`) and assert they got passed the right things rather than implicitly test that logic a second time in the client tests.

Finally, adding integration tests for the `LivyBatch` object you've added would be strongly valuable. As you can see from tests/test_integration.py, these tests run the code against an actual livy server in a realistic-ish setting. The batch tests should be simpler than the session ones, as you don't have all the dataframe export logic to test, but it would be good to see a test that causes some code to be run on the server side, and some expected output appear in the logs.

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@ dist/
 livy.egg-info/
 .pytest_cache/
 .tox/
+.idea/
```
Comment (acroz) [inline .gitignore]: I generally prefer not to include editor-specific ignores. Can you please remove this and add it instead to [a global ignores file](https://acroz.dev/2019/07/14/global-git-ignore/)?

File: livy/batch.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        self.echo = echo
+        self.check = check
+        self.batch_id: Optional[int] = None
+
+    def __enter__(self) -> "LivyBatch":
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback) -> None:
+        self.kill()
```
Comment (acroz) [inline livy/batch.py]: How do you expect the `LivyBatch` object to be used?

The usage of `LivySession` as a context manager enables it to be automatically closed when the code block exits, but I'm not sure this makes sense for a non-interactive batch session.

Would it make more sense to drop the context manager behaviour and have something that works like:

```python
batch = LivyBatch(...)
batch.start()  # Actually creates the batch in Livy
batch.wait()  # Waits for a state indicating the batch is finished (either successfully or through an error)
batch.logs  # A `@property` that fetches the latest log lines
```

Keeping `batch.kill()` seems like a valuable method available to the user if their batch gets stuck.

I'd suggest that  `start` does not wait for the batch to actually begin - this was necessary for the interactive session in `LivySession` but does not seem so valuable here.

**Comment by acroz** [comment]
Oh, and can you also please add yourself to the `AUTHORS` file?

**Comment by JustEdro** [comment]
Thank you for describing that thoroughly, fix will take some time.

File: livy/batch.py (hunk 2 above)
Comment (JustEdro) [reply livy/batch.py]: Makes sense, thanks!
Also, I kept the `state` property to leave the ability for the program to perform asynchronously.

Comment (acroz) [reply livy/batch.py]: > Also, I kept the `state` property to leave the ability for the program to perform asynchronously.

Sounds like a good idea!

**Comment by dineshdharme** [comment]
@acroz , When will the Batches endpoint be merged in ? Our requirement currently needs to use batches endpoint. Thanks!

**Comment by JustEdro** [comment]
@dineshdharme Hi, that's my bad. I've got no time to finish the tests yet. 
If you can finish them your help will be very appreciated

**Comment by JustEdro** [comment]
@acroz, could you please rebuild the docker image `acroz/livy` so that I could finish the integration tests? 
https://issues.apache.org/jira/browse/SPARK-21917 blocks me from that now.

**Comment by yustoris** [comment]
Any progress in this? I'm very looking forward to this feature 🙂.

**Comment by JustEdro** [comment]
I have finished the tests but am unable to run the whole testing because the maintainer does not respond. :(

**Comment by acroz** [comment]
Hi both, sorry for the delay on this - I did not have time to address
Eduard's request at the time and let it slip through the net. I'm away
without my laptop but will look into this next week.

**Comment by ketgo** [comment]
@JustEdro You can use the following docker file to create Livy+Spark image for testing:
```Dockerfile
FROM debian:stretch
MAINTAINER Ketan Goyal

RUN apt-get update \
 && apt-get install -y locales \
 && dpkg-reconfigure -f noninteractive locales \
 && locale-gen C.UTF-8 \
 && /usr/sbin/update-locale LANG=C.UTF-8 \
 && echo "en_US.UTF-8 UTF-8" >> /etc/locale.gen \
 && locale-gen \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*

# Users with other locales should set this in their derivative image
ENV LANG en_US.UTF-8
ENV LANGUAGE en_US:en
ENV LC_ALL en_US.UTF-8

RUN apt-get update \
 && apt-get install -y curl unzip \
    python3 python3-setuptools \
 && ln -s /usr/bin/python3 /usr/bin/python \
 && easy_install3 pip py4j \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*

# http://blog.stuart.axelbrooke.com/python-3-on-spark-return-of-the-pythonhashseed
ENV PYTHONHASHSEED 0
ENV PYTHONIOENCODING UTF-8
ENV PIP_DISABLE_PIP_VERSION_CHECK 1

# JAVA
RUN apt-get update \
 && apt-get install -y openjdk-8-jre \
 && apt-get clean \
 && rm -rf /var/lib/apt/lists/*

# HADOOP
ENV HADOOP_VERSION 3.0.0
ENV HADOOP_HOME /usr/hadoop-$HADOOP_VERSION
ENV HADOOP_CONF_DIR=$HADOOP_HOME/etc/hadoop
ENV PATH $PATH:$HADOOP_HOME/bin
RUN curl -sL --retry 3 \
  "http://archive.apache.org/dist/hadoop/common/hadoop-$HADOOP_VERSION/hadoop-$HADOOP_VERSION.tar.gz" \
  | gunzip \
  | tar -x -C /usr/ \
 && rm -rf $HADOOP_HOME/share/doc \
 && chown -R root:root $HADOOP_HOME

# SPARK
ENV SPARK_VERSION 2.4.1
ENV SPARK_PACKAGE spark-${SPARK_VERSION}-bin-without-hadoop
ENV SPARK_HOME /usr/spark-${SPARK_VERSION}
ENV SPARK_DIST_CLASSPATH="$HADOOP_HOME/etc/hadoop/*:$HADOOP_HOME/share/hadoop/common/lib/*:$HADOOP_HOME/share/hadoop/common/*:$HADOOP_HOME/share/hadoop/hdfs/*:$HADOOP_HOME/share/hadoop/hdfs/lib/*:$HADOOP_HOME/share/hadoop/hdfs/*:$HADOOP_HOME/share/hadoop/yarn/lib/*:$HADOOP_HOME/share/hadoop/yarn/*:$HADOOP_HOME/share/hadoop/mapreduce/lib/*:$HADOOP_HOME/share/hadoop/mapreduce/*:$HADOOP_HOME/share/hadoop/tools/lib/*"
ENV PATH $PATH:${SPARK_HOME}/bin
RUN curl -sL --retry 3 \
  "https://archive.apache.org/dist/spark/spark-${SPARK_VERSION}/${SPARK_PACKAGE}.tgz" \
  | gunzip \
  | tar x -C /usr/ \
 && mv /usr/$SPARK_PACKAGE $SPARK_HOME \
 && chown -R root:root $SPARK_HOME

# Livy
ENV LIVY_VERSION 0.7.0
ENV LIVY_PACKAGE apache-livy-${LIVY_VERSION}-incubating-bin
ENV LIVY_HOME /usr/livy-${LIVY_VERSION}
ENV PATH $PATH:${LIVY_HOME}/bin
RUN apt-get update \
 && apt-get install -y procps
RUN curl -sL --retry 3 \
  "http://mirror.cogentco.com/pub/apache/incubator/livy/${LIVY_VERSION}-incubating/${LIVY_PACKAGE}.zip" > ${LIVY_PACKAGE}.zip \
  && unzip -d /usr/ ${LIVY_PACKAGE}.zip \
  && rm ${LIVY_PACKAGE}.zip \
  && mv /usr/$LIVY_PACKAGE $LIVY_HOME \
  && mkdir ${LIVY_HOME}/logs \
  && chown -R root:root $LIVY_HOME

CMD ["livy-server"]
```

I have a need for the batches API and am really looking forward to this pull request being merged. 

Hope that helps!

**Comment by acroz** [comment]
@ketgo, when running tests against your Docker setup, things run fine on Scala and R, but I'm having issues with PySpark. Specifically, I get the following error, which causes the session to crash:

```
Exception in thread "Thread-25" java.lang.NoClassDefFoundError: org/apache/spark/sql/hive/HiveContext
```

Is this something you've encountered? I'm able to run and use the PySpark shell directly by shelling into the running container without issues.

**Comment by ketgo** [comment]
Hi @acroz, I have not run into this issue before. Can you point to the test or code for which you are observing the error. I would like to reproduce it for debugging. 

My initial guess would be it seems that some of the hive libs are missing or there is a version mismatch.

**Comment by acroz** [comment]
Hey @JustEdro - I finally managed to muster the motivation to fix the integration tests. There was some hairy stuff going on with the Python interpreter failing as a particular Java class was not available.

Anyway, I've merged the latest master with fixed tests on to your branch! Can I help at all to fix any of the issues revealed?

**Comment by JustEdro** [comment]
Hi @acroz, sorry for the belated reply

I've run the tests on the updated Docker image, and everything works like a charm.
The endpoint itself works well, so I think these changes are ready to be merged to master if you don't mind.

**Comment by jespercedergren** [comment]
> @ketgo, when running tests against your Docker setup, things run fine on Scala and R, but I'm having issues with PySpark. Specifically, I get the following error, which causes the session to crash:
> 
> ```
> Exception in thread "Thread-25" java.lang.NoClassDefFoundError: org/apache/spark/sql/hive/HiveContext
> ```
> 
> Is this something you've encountered? I'm able to run and use the PySpark shell directly by shelling into the running container without issues.



> @ketgo, when running tests against your Docker setup, things run fine on Scala and R, but I'm having issues with PySpark. Specifically, I get the following error, which causes the session to crash:
> 
> ```
> Exception in thread "Thread-25" java.lang.NoClassDefFoundError: org/apache/spark/sql/hive/HiveContext
> ```
> 
> Is this something you've encountered? I'm able to run and use the PySpark shell directly by shelling into the running container without issues.

Hi @acroz, 
I encounter the same thing as you. It throws me the same exception when I post single statements when I used the same Dockerfile provided by @ketgo above. How did you solve this? 
Help would be much appreciated!

**Comment by acroz** [comment]
Hi @jespercedergren, I added a command to the Dockerfile to install
`spark-hive`, which provides the missing dependency. You can see it here:

https://github.com/acroz/pylivy-test-server/blob/17d690bf0d1dacafb1a9413e49ccdedef190a07a/Dockerfile#L63-L66

Hope this helps!

**Comment by jespercedergren** [comment]
Thanks @acroz for you quick response!

**Pull request merged**